icechunk/src/format/snapshot.rs

```diff
@@ -49,6 +49,55 @@ pub struct ZarrArrayMetadata {
     pub dimension_names: Option<DimensionNames>,
 }
 
+impl ZarrArrayMetadata {
+
+    /// Returns an iterator over the maximum permitted chunk indices for the array.
+    ///
+    /// This function calculates the maximum chunk indices based on the shape of the array
+    /// and the chunk shape.
+    ///
+    /// # Returns
+    ///
+    /// An iterator over the maximum permitted chunk indices.
+    fn max_chunk_indices_permitted(&self) -> impl Iterator<Item = u64> + '_ {
```

Comment:
I think chunk indices are u32:
```rust
pub struct ChunkIndices(pub Vec<u32>);
```

At the beginning of this method, it could be a good place to:
```rust
debug_assert_eq!(self.shape.len(), self.chunk_shape.0.len())
```
or similar.

Reply:
Pushed a new commit to address this. I ended up returning a `ChunkIndices` type from this using `collect()` on the iterator and type casting the calculation result to u32. Not sure if there is a performance consideration with using collect prior to returning, but it seemed to make sense that the type returned be `ChunkIndices`. This way the actual validation function compares two ChunkIndices.

Shape and chunk shape were both u64, and I assume chunk indices are u32 since they will always be much smaller values. 

Also added the debug assert, which is causing tests in `repository.rs` to fail. I noticed this is because the chunk shape initialized is a single value, even when the shape is multiple. Do we assume that if a single chunk shape value is present, it applies to all dimensions? Or should this fail?

Reply:
> I ended up returning a ChunkIndices

perfect!

> I assume chunk indices are u32

yes, you cand valitade looking [here](https://github.com/earth-mover/icechunk/blob/ee982e714d25e1a5681f216141fe95a88f5b2f40/icechunk/src/format/mod.rs#L141)

> chunk shape initialized is a single value, even when the shape is multiple.

this is probably a bad test, we should fix it to match. We need to add more validations to the metadata to avoid this kind of thing, but it's not critical because zarr also validates.

> Do we assume that if a single chunk shape value is present, it applies to all dimensions

No, I don't think so, probably just a bad test, please change it to match.

Reply:
Thanks for answers! I will return to this PR next week after the holiday.